lib/core/checker/checker.cc

```diff
@@ -10,6 +19,263 @@
 
 namespace dbuf::checker {
 
+struct NameResolutionChecker {
+  template <typename T>
+  bool operator()(const T &) const {
+    return false;
+  }
+
+  bool operator()(const ast::AST &ast) {
+    AddScope();
+    AddGlobalNames(ast);
+
+    constructors_fields_ = GetConstructorFields(ast);
+
+    if (!(*this)(ast.messages)) {
+      return false;
+    }
+
+    return (*this)(ast.enums);
+  }
+
+  bool operator()(const std::unordered_map<InternedString, ast::Enum> &enums) {
+    return !std::ranges::any_of(enums.begin(), enums.end(), [this](auto &ast_enum) {
+      return !(*this)(ast_enum.second);
+    });
+  }
+
+  bool operator()(const ast::Enum &ast_enum) {
+    AddScope();
+    if (!IsInScope(ast_enum.name)) {
+      return false;
+    }
+    if (!(*this)(ast_enum.type_dependencies)) {
+      return false;
+    }
+    if (!(*this)(ast_enum.pattern_mapping)) {
+      return false;
+    }
+    DeleteScope();
+
+    return true;
+  }
+
+  bool operator()(const std::vector<ast::Enum::Rule> &rules) {
+    return !std::ranges::any_of(rules.begin(), rules.end(), [this](auto &rule) {
+      return !(*this)(rule);
+    });
+  }
+
+  bool operator()(const ast::Enum::Rule &rule) {
+    AddScope();
+
+    isShadowing_ = true;
+
+    if (!(*this)(rule.inputs)) {
+      return false;
+    }
+
+    isShadowing_ = false;
```

Comment:
Isn't the better place for this flag setting in the method that checks the inputs? Set at the beginning and unset at the end. I think we should unset it before returning because it might impact later checks when we move to collecting errors instead of doing it one by one.